libzim/libzim.pyx

```diff
@@ -0,0 +1,286 @@
+cimport libzim
+cimport cpython.ref as cpy_ref
+from cython.operator import dereference
+
+from libc.stdint cimport uint32_t, uint64_t
+from libcpp.string cimport string
+from libcpp cimport bool
+from libcpp.memory cimport shared_ptr, make_shared
+
+import datetime
+
+from collections import defaultdict
+
+#########################
+#       ZimBlob         #
+#########################
+
+cdef class ZimBlob:
+    cdef Blob* c_blob
+
+    def __init__(self, content):
+
+        if isinstance(content, str):
+            ref_content = content.encode('UTF-8')
+        else:
+            ref_content = content
+
+        self.c_blob = new Blob(<char *> ref_content, len(ref_content))
+
+    def __dealloc__(self):
+        if self.c_blob != NULL:
+            del self.c_blob
+
+
+#########################
+#       ZimArticle      #
+#########################
+
+cdef class ZimArticle:
+    cdef ZimArticleWrapper* c_article
+
+    def __init__(self):
+        self.c_article = new ZimArticleWrapper(<cpy_ref.PyObject*>self)
+    
+    def get_url(self):
+        raise NotImplementedError
+
+    def get_title(self):
+        raise NotImplementedError
+    
+    def is_redirect(self):
+        raise NotImplementedError
+
+    def get_mime_type(self):
+        raise NotImplementedError
+
+    def get_filename(self):
+        raise NotImplementedError
+
+    def should_compress(self):
+        raise NotImplementedError
+
+    def should_index(self):
+        raise NotImplementedError
+
+    def redirect_url(self):
+        raise NotImplementedError
+
+    def get_data(self):
+        raise NotImplementedError
+
+    @property
+    def mimetype(self):
+        return self.c_article.getMimeType().decode('UTF-8')
+    
+    @property
+    def content(self):
+        blob = self.c_article.getData()
+        content =  blob.data()[:blob.size()]
+        try:
+            return content.decode('UTF-8')
+        except UnicodeDecodeError:
+            return content
+
+    # This changes with implementation
+    @property
+    def can_write(self):
+        raise NotImplementedError
+
+
+
+#------- ZimArticle pure virtual methods --------
+cdef public api:
+    string string_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a string"""
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
+            value = func()
+            return value.encode('UTF-8')
+    
+    Blob blob_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a Blob"""
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        cdef ZimBlob blob = ZimBlob(b'\x00')
```

Comment:
Please do not create a ZimBlob with a dummy bytes.
It is not necessary (it ends by allocating a new `zim::Blob`) and it lead the reader to wrongly assume the bytes is useful.

Reply:
Agree with you, but I can't declare a cdef inside conditionals (cython limitation) it should be at  the start, I need to have an "empty object"  to capture the returned Blob.  Maybe you can help me figure out a less elegant way to do it while respecting Cython's requirement.

```python3
Blob blob_cy_call_fct(void *ptr, string method, int *error):
        """Lookup and execute a pure virtual method on ZimArticle returning a Blob"""
        cdef ZimArticle art = <ZimArticle>(ptr)
        cdef ZimBlob blob = ZimBlob(b'\x00')
        try:
            func = getattr(art, method.decode('UTF-8'))
        except AttributeError:
            error[0] = 1
            raise 
        else:
            error[0] = 0
            blob = func()
            return dereference(blob.c_blob) 
```

Reply:
Fix https://github.com/openzim/python-libzim/pull/3/commits/a8c5bafd4f5ea761190844d1097ba73d55033c5f